LFS-747 / LFS-802: Multiple choice questions should support the "None of the above" option

OrdiNeu opened the pull request:
[Jira link for the base functionality](https://phenotips.atlassian.net/browse/LFS-747) 

[Jira link for the questionnaire changes](https://phenotips.atlassian.net/browse/LFS-802)

**Test branch:** Running the multiple choice in `cardiac_rehab` runmode (`java -jar distribution/target/lfs-0.1-SNAPSHOT.jar -Dsling.run.modes=nolfs,cardiac_rehab,dev`),
- create a form for `6MWT` and check the behaviour of `Reason for pausing during the test`. The NA option should deselect everything, including user-input ones.
- create a form for `Baseline Health Information`, and check the behaviour of `Diseases of the musculoskeletal system or connective tissue: check all that apply`. The "None of the above" option should deselect everything except for anything you enter in the input box.

This adds another option to an lfs:AnswerOption that allows you to make an answer option a "none of the above". This isn't yet added to the editor.

OrdiNeu committed:
LFS-747: Multiple choice questions should support the "None of the above" option

Added the "none of the above" behaviour to an answerOption.

sashaandjic reviewed (approved):
tested

sashaandjic commented:
Will the editor changes be included in this PR?

sashaandjic commented:
None of the above works well
![image](https://user-images.githubusercontent.com/23243244/102553489-ec5e3f80-4090-11eb-8a83-1c6e8479ffbc.png)

Other options are shown as IDs nit as description:
![image](https://user-images.githubusercontent.com/23243244/102553525-fc761f00-4090-11eb-805d-6a7fd4efda85.png)

marta- commented:
> Will the editor changes be included in this PR?

No. Editor support for answer options is already very limited (only values are supported, not labels). We will expand editor support in another task.

k-pace commented:
In cardiac, 0-5ND form, first question contains 'None' (instead of 'None of the above') which does not clear responses on selection
![image](https://user-images.githubusercontent.com/33154316/102628082-28d47e80-4117-11eb-9f84-a3adcc16722e.png)

k-pace commented:
tested

marta- reviewed (approved):
Looks great

OrdiNeu committed:
LFS-747: Multiple choice questions should support the "None of the
above" option

Remove any user-input questions when selecting a "None of the above"
option

OrdiNeu committed:
LFS-802: Enable "none of the above" functionality in cardiac rehab
questionnaires

OrdiNeu committed:
LFS-747: Multiple choice questions should support the "None of the
above" option

Also add a "na" option, which clears every input

OrdiNeu committed:
LFS-802: Enable "none of the above" functionality in cardiac rehab
questionnaires

Add the N/A field functionality.

marta- reviewed (approved):
New code changes approved, light testing done.

sashaandjic reviewed (approved):
tested - all good

OrdiNeu committed:
LFS-747: Multiple choice questions should support the "None of the
above" option

Fix a bug where the internal "selected" state would exclude user-defined
options in a "none of the above" answer.

k-pace commented:
tested - looks good. None of the above and free text values both stored

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -83,6 +83,14 @@
+  - noneOfTheAbove (boolean)
+
+  // If set, selecting this answer in a multiple choice question will remove _all other options, including user-entered ones_
+  // (as in a "not applicable" option) In addition, selecting any other answer will unselect this option
```
```suggestion
  // (as in a "not applicable" option). In addition, selecting any other answer will unselect this option.
```

sdumitriu commented on modules/data-entry/src/main/resources/SLING-INF/nodetypes/dataentry.cnd:
```diff
@@ -83,6 +83,14 @@
   - value (string) mandatory
 
+  // If set, selecting this answer in a multiple choice question will remove _all other non-user-entered options_
+  // (as in a "none of the above" option). In addition, selecting any other answer will unselect this option
```
```suggestion
  // (as in a "none of the above" option). In addition, selecting any other answer will unselect this option.
```

sdumitriu committed:
[misc] Add dots where needed

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/MultipleChoice.jsx:
```diff
@@ -105,6 +112,29 @@ function MultipleChoice(props) {
+        let defaultOptions = defaults.filter(option => option[IS_DEFAULT_POS]).map((option) => option[VALUE_POS]);
+        let newSelection = old.slice().filter((option) => !defaultOptions.includes(option[VALUE_POS]));
+        newSelection.push([name, id]);
+        console.log(newSelection);
```
Should have removed this...
